projects/cm_mcu/clocksynth.c

```diff
@@ -153,3 +154,92 @@ int load_clock()
+  // ensure that the device is in the right range 0-5
+  if (device < 0 || device > 4)
+    return;
+
+  // grab the semaphore to ensure unique access to I2C controller
+  if (acquireI2CSemaphore(i2c2_sem) == pdFAIL) {
+    log_warn(LOG_SERVICE, "could not get semaphore in time\r\n");
+    return;
+  }
+  // extract info about device
+  uint8_t mux_addr, mux_bit, dev_addr;
+  uint16_t eeprom_progname_reg;
+  // In Rev2 device 0 and devices 1-5 are different and hence are stored in different arrays
+  // for monitoring purposes
+  if (device == 0) {
+    mux_addr = clkr0a_moni2c_addrs[0].mux_addr;
+    mux_bit = clkr0a_moni2c_addrs[0].mux_bit;
+    dev_addr = clkr0a_moni2c_addrs[0].dev_addr;
+    eeprom_progname_reg = clkr0a_moni2c_addrs[0].eeprom_progname_reg;
+  }
+  else {
+    mux_addr = clk_moni2c_addrs[device - 1].mux_addr;
+    mux_bit = clk_moni2c_addrs[device - 1].mux_bit;
+    dev_addr = clk_moni2c_addrs[device - 1].dev_addr;
+    eeprom_progname_reg = clk_moni2c_addrs[device - 1].eeprom_progname_reg;
+  }
+  // set mux bit
+  int status = apollo_i2c_ctl_w(CLOCK_I2C_DEV, mux_addr, 1, 1 << mux_bit);
+  if (status != 0) {
+    log_debug(LOG_I2C, "mux write stat=%d\r\n", status); // can't return due to semaphore
+  }
+  else {
+    // set the page
+    uint8_t page = (CLOCK_PROGNAME_REG_ADDR_START >> 8) & 0xFF;
+    status = apollo_i2c_ctl_reg_w(CLOCK_I2C_DEV, dev_addr, 1,
+                                  CLOCK_CHANGEPAGE_REG_ADDR, 1, page);
+
+    // now read out the six bytes of data in two reads
+    const uint8_t reg = (CLOCK_PROGNAME_REG_ADDR_START)&0xFF;
+    uint16_t init_postamble_page = 32 * (device + 1) - 1;
+
+    uint32_t PreambleList_row; // the size of preamble list in a clock config file store at the end of the last eeprom page of a clock
+    status += apollo_i2c_ctl_reg_r(CLOCK_I2C_DEV, CLOCK_I2C_EEPROM_ADDR, 2, (init_postamble_page << 8) + 0x007C, 1, &PreambleList_row);
+    uint32_t RegisterList_row; // the size of register list in a clock config file store at the end of the last eeprom page of a clock
+    status += apollo_i2c_ctl_reg_r(CLOCK_I2C_DEV, CLOCK_I2C_EEPROM_ADDR, 2, (init_postamble_page << 8) + 0x007D, 2, &RegisterList_row);
+    uint32_t PostambleList_row; // the size of postamble list in a clock config file store at the end of the last eeprom page of a clock
+    status += apollo_i2c_ctl_reg_r(CLOCK_I2C_DEV, CLOCK_I2C_EEPROM_ADDR, 2, (init_postamble_page << 8) + 0x007F, 1, &PostambleList_row);
+
+    uint32_t eepromdata[2];
+    if (PreambleList_row == 0xff && RegisterList_row == 0xffff && PostambleList_row == 0xff) { // check if a clock has been programmed or not from a set of three registers
+      eepromdata[0] = 0x58;                                                                    // supposed to be an "X" for an unprogrammed clock
+      eepromdata[1] = 0x00;
+    }
+    else {
+      eepromdata[0] = 0UL;
+      eepromdata[1] = 0UL;
+      uint32_t tempdata[2];
+      for (int i = 0; i < 4; ++i) {
```

Comment:
if we are reading 6 memory locations of a byte each why do we need 8 I2C transactions?

Reply:
I maybe able to read one I2C transaction with 12 bytes and apply a bit mask for eeprom[0]([1]) but that is more than the 4-byte limit.

Reply:
that makes sense now, thanks!